nilearn/datasets/tests/test_func.py

```diff
@@ -0,0 +1,345 @@
+"""
+Test the datasets module
+"""
+# Author: Alexandre Abraham
+# License: simplified BSD
+
+import os
+import shutil
+import numpy as np
+from tempfile import mkdtemp
+
+from nose import with_setup
+from nose.tools import assert_true, assert_equal, assert_raises
+
+
+from nilearn import datasets
+from nilearn._utils.testing import (mock_request, wrap_chunk_read_,
+                                    FetchFilesMock)
+
+from nilearn._utils.compat import _basestring
+
+
+currdir = os.path.dirname(os.path.abspath(__file__))
+datadir = os.path.join(currdir, 'data')
+tmpdir = None
+url_request = None
+file_mock = None
+
+
+def setup_tmpdata():
+    # create temporary dir
+    global tmpdir
+    tmpdir = mkdtemp()
+
+
+def setup_mock():
+    global url_request
+    url_request = mock_request()
+    datasets.utils._urllib.request = url_request
+    datasets.utils._chunk_read_ = wrap_chunk_read_(datasets.utils._chunk_read_)
+    global file_mock
+    file_mock = FetchFilesMock()
+    datasets.func._fetch_files = file_mock  # overwrite the actual function
+
+
+def teardown_tmpdata():
+    # remove temporary dir
+    global tmpdir
+    if tmpdir is not None:
+        shutil.rmtree(tmpdir)
+
+
+@with_setup(setup_mock)
```

Comment:
The solution is to have a closing functions for the test that put the original modules back instead of the mock. This is cleaner and safer.

Could be:

```python
original_fetch_files = None
original_url_request = None
original_chunk_read = None
mock_fetch_files = None
mock_url_request = None
mock_chunk_read = None

def setup_mock():
    global original_url_request
    global mock_url_request
    mock_url_request = mock_request()
    original_url_request = datasets.utils._urllib.request
    datasets.utils._urllib.request = mock_url_request

    global original_chunk_read
    global mock_chunk_read
    mock_chunk_read = wrap_chunk_read_(datasets.utils._chunk_read_)
    original_chunk_read = datasets.utils._chunk_read_
    datasets.utils._chunk_read_ = mock_chunk_read

    global original_fetch_files
    global mock_fetch_files
    mock_fetch_files = FetchFilesMock()
    original_fetch_files = datasets.func._fetch_files
    datasets.func._fetch_files = mock_fetch_files

def teardown_mock():
    global original_url_request
    datasets.utils._urllib.request = original_url_request

    global original_chunk_read
    datasets.utils._chunk_read_ = original_chunk_read

    global original_fetch_files
    datasets.func._fetch_files = original_fetch_files
```